src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,286 @@
+    }
+  }
+
+  out.psi_inverse = out.psi.cwiseInverse().asDiagonal();
+
+  return out;
+}
+
+void MFAHierarchy::initialize_state() {
+  std::cout << "init state" << std::endl;
+  state.mu = hypers->mutilde;
+  state.psi = hypers->beta / (hypers->alpha0 + 1.);
+  state.eta = Eigen::MatrixXd::Zero(card, hypers->q);
+  state.lambda = Eigen::MatrixXd::Zero(dim, hypers->q);
+  state.psi_inverse = state.psi.cwiseInverse().asDiagonal();
+}
+
+void MFAHierarchy::initialize_hypers() {
+  if (prior->has_fixed_values()) {
+    // Set values
+    hypers->mutilde = bayesmix::to_eigen(prior->fixed_values().mutilde());
+    dim = hypers->mutilde.size();
+    hypers->beta = bayesmix::to_eigen(prior->fixed_values().beta());
+    hypers->phi = prior->fixed_values().phi();
+    hypers->alpha0 = prior->fixed_values().alpha0();
+    hypers->q = prior->fixed_values().q();
+
+    // Automatic initialization
+    if (dim == 0) {
+      hypers->mutilde = dataset_ptr->colwise().mean();
+      dim = hypers->mutilde.size();
+    }
+    if (hypers->beta.size() == 0) {
+      Eigen::MatrixXd centered =
+          dataset_ptr->rowwise() - dataset_ptr->colwise().mean();
+      auto cov_llt = ((centered.transpose() * centered) /
+                      double(dataset_ptr->rows() - 1.))
+                         .llt();
+      Eigen::MatrixXd precision_matrix(
+          cov_llt.solve(Eigen::MatrixXd::Identity(dim, dim)));
+      hypers->beta =
+          (hypers->alpha0 - 1) * precision_matrix.diagonal().cwiseInverse();
+      if (hypers->alpha0 == 1) {
+        throw std::invalid_argument(
+            "Scale parameter must be different than 1 when automatic "
+            "initialization is used");
+      }
+    }
+    // Check validity
+    if (dim != hypers->beta.rows()) {
+      throw std::invalid_argument(
+          "Hyperparameters dimensions are not consistent");
+    }
+    for (size_t j = 0; j < dim; j++) {
+      if (hypers->beta[j] <= 0) {
+        throw std::invalid_argument("Shape parameter must be > 0");
+      }
+    }
+    if (hypers->alpha0 <= 0) {
+      throw std::invalid_argument("Scale parameter must be > 0");
+    }
+    if (hypers->phi <= 0) {  // TODO check
+      throw std::invalid_argument("Diffusion parameter must be > 0");
+    }
+    if (hypers->q <= 0) {
+      throw std::invalid_argument("Number of factors must be > 0");
+    }
+  }
+
+  else {
+    throw std::invalid_argument("Unrecognized hierarchy prior");
+  }
+}
+
+void MFAHierarchy::update_hypers(
+    const std::vector<bayesmix::AlgorithmState::ClusterState>& states) {
+  auto& rng = bayesmix::Rng::Instance().get();
+  if (prior->has_fixed_values()) {
+    return;
+  }
+
+  else {
+    throw std::invalid_argument("Unrecognized hierarchy prior");
+  }
+}
+
+void MFAHierarchy::update_summary_statistics(const Eigen::RowVectorXd& datum,
+                                             bool add) {
+  if (add) {
+    data_sum += datum;
+  } else {
+    data_sum -= datum;
+  }
+}
```

Comment:
This should probably be the biggest improvement with the new approach, since we don't need to scan the whole dataset to add or remove points at each iteration anymore.